stan/math/prim/scal/fun/grad_reg_lower_inc_gamma.hpp

```diff
@@ -0,0 +1,189 @@
+ * function for small to moderate values of $z$ using the
+ * approximation:
+ *
+ * \f[
+ *  \frac{\gamma(a,z)}{\Gamma(a)}=z^a e^-z
+ *    \sum_n=0^\infty \frac{z^n}{\Gamma(a+n+1)}
+ * \f]
+ *
+ * We write the derivative in the form:
+ *
+ * \f[
+ *   \frac{d\gamma(a,z)\Gamma(a)}{da} = \frac{\log z}{e^z}
+ *     \sum_n=0^\infty \frac{z^{a+n}}{\Gamma(a+n+1)}
+ *   - \frac{1}{e^z}
+ *     \sum_n=0^\infty \frac{z^{a+n}}{\Gamma(a+n+1)}\psi^0(a+n+1)
+ * \f]
+ *
+ * This caclulation is sufficiently accurate for small $a$ and
+ * small $z$.  For larger values and $a$ and $z$ we use it in its
+ * log form:
+ *
+ * \f[
+ *   \frac{d \gamma(a,z)\Gamma(a)}{da} = \frac{\log z}{e^z}
+ *     \sum_n=0^\infty \exp[(a+n)\log z - \log\Gamma(a+n+1)]
+ *   - \sum_n=0^\infty \exp[(a+n)\log z - \log\Gamma(a+n+1) +
+ *       \log\psi^0(a+n+1)]
+ * \f]
+ *
+ * For large $z$, Gauschi recommends using the upper incomplete
+ * Gamma instead and the negative of its derivative turns out to be
+ * more stable and accurate for larger $z$ and for some combinations
+ * of $a$ and $z$. This is a log-scale implementation of the
+ * derivative of the formulation suggested by Gauschi (1979). For
+ * some values it defers to the negative of the gradient
+ * for the gamma_q function. This is based on the suggestion by Gauschi
+ * (1979) that for large values of $z$ it is better to
+ * carry out calculations using the upper incomplete Gamma function.
+ *
+ * Branching for choice of implementation for the lower incomplete
+ * regularized gamma function gradient. The derivative based on
+ * Gautschi's formulation appears to be sufficiently accurate
+ * everywhere except for large z and small to moderate a. The
+ * intersection between the two regions is a radius 12 quarter circle
+ * centered at a=0, z=30 although both implementations are
+ * satisfactory near the intersection.
+ *
+ * Some limits that could be treated, e.g., infinite z should
+ * return tgamma(a) * digamma(a), throw instead to match the behavior of,
+ * e.g., boost::math::gamma_p
+ *
+ * @tparam T1 type of a
+ * @tparam T2 type of z
+ * @param[in] a shared with complete Gamma
+ * @param[in] z value to integrate up to
+ * @param[in] precision series terminates when increment falls below
+ * this value.
+ * @param[in] max_steps number of terms to sum before throwing
+ * @throw std::domain_error if the series does not converge to
+ * requested precision before max_steps.
+ *
+ */
+template <typename T1, typename T2>
+typename return_type<T1, T2>::type grad_reg_lower_inc_gamma(
+    const T1& a, const T2& z, double precision = 1e-10, int max_steps = 1e5) {
+  using std::exp;
+  using std::log;
+  using std::pow;
+  typedef typename return_type<T1, T2>::type TP;
+
+  if (is_nan(a))
+    return std::numeric_limits<TP>::quiet_NaN();
+  if (is_nan(z))
+    return std::numeric_limits<TP>::quiet_NaN();
+  if (a > 0 && is_inf(a))
+    domain_error("d_lower_reg_inc_gamma_da", "a", a, "The value ",
```

Comment:
the function passed into the `check_` function should be the name of this function, `grad_reg_lower_inc_gamma`. That's both for consistency and for the user guiding us to the right function if this pops up.